src/vs/workbench/contrib/chat/browser/chatPasteProviders.ts

```diff
@@ -136,13 +152,126 @@ export function isImage(array: Uint8Array): boolean {
 	);
 }
 
+export class CopyTextProvider implements DocumentPasteEditProvider {
+
+	static readonly id = 'text';
+	static readonly kind = new HierarchicalKind('text.plain');
+
+	readonly id = PasteTextProvider.id;
+	readonly kind = PasteTextProvider.kind;
+
+	readonly copyMimeTypes = [Mimes.text, 'vscode-editor-data'];
+
+	async prepareDocumentPaste(model: ITextModel, ranges: readonly IRange[], dataTransfer: IReadonlyVSDataTransfer, token: CancellationToken): Promise<undefined | IReadonlyVSDataTransfer> {
+		if (model.uri.scheme === ChatInputPart.INPUT_SCHEME) {
+			return;
+		}
+		const customDataTransfer = new VSDataTransfer();
+		const rangesString = JSON.stringify({ ranges: ranges[0], uri: model.uri.toString() });
+		customDataTransfer.append('additional-editor-data', createStringDataTransferItem(rangesString));
+		return customDataTransfer;
+	}
+}
+
+export class PasteTextProvider implements DocumentPasteEditProvider {
+
+	static readonly id = 'text';
+	static readonly kind = new HierarchicalKind('text.plain');
+
+	readonly id = PasteTextProvider.id;
+	readonly kind = PasteTextProvider.kind;
+
+	readonly pasteMimeTypes = [Mimes.text, 'vscode-editor-data', 'additional-editor-data'];
+
+	constructor(
+		private readonly chatWidgetService: IChatWidgetService
+	) { }
+
+
+	async provideDocumentPasteEdits(model: ITextModel, ranges: readonly IRange[], dataTransfer: IReadonlyVSDataTransfer, context: DocumentPasteContext, token: CancellationToken): Promise<DocumentPasteEditsSession | undefined> {
+		if (model.uri.scheme !== ChatInputPart.INPUT_SCHEME) {
+			return;
+		}
+		const text = dataTransfer.get(Mimes.text);
+		const editorData = dataTransfer.get('vscode-editor-data');
+		const additionalEditorData = dataTransfer.get('additional-editor-data');
+
+		if (!editorData || !text || !additionalEditorData) {
+			return;
+		}
+
+		const textdata = await text.asString();
+		const metadata = JSON.parse(await editorData.asString());
+		const additionalData = JSON.parse(await additionalEditorData.asString());
+
+		const widget = this.chatWidgetService.getWidgetByInputUri(model.uri);
+		if (!widget) {
+			return;
+		}
+
+		const copiedContext = await getCopiedContext(textdata, additionalData.uri, metadata.mode, additionalData.ranges);
+
+		if (token.isCancellationRequested || !copiedContext) {
+			return;
+		}
+
+		const currentContextIds = widget.attachmentModel.getAttachmentIDs();
+		if (currentContextIds.has(copiedContext.id)) {
+			return;
+		}
+
+		const customEdit = {
+			resource: model.uri,
+			variable: copiedContext,
+			undo: () => { widget.attachmentModel.delete(copiedContext.id); },
```

Comment:
You may need to try getting the widget in these callbacks just in case it the widget has been invalidated after the paste has run